lib/fhir/common.js

```diff
@@ -119,6 +110,196 @@ module.exports = exports = (mongo) => {
     throw new Error(`Unknown ID generator '${generator}'. Valid options are 'uuidv1' or 'uuidv4'`)
   }
 
+  function resolveReference (reference, callback) {
+    const matches = REFERENCE_REGEX.exec(reference)
+    if (!matches) {
+      return callback(new Error(`Invalid resource reference "${reference}"`))
+    }
+    mongo.getDB((err, db) => {
+      if (err) {
+        return callback(err)
+      }
+      const resourceType = matches[4]
+      const resourceId = matches[5]
+      const versionId = matches[7]
+      if (versionId) {
+        db
+          .collection(`${resourceType}_history`)
+          .findOne({id: resourceId, 'meta.versionId': versionId}, callback)
+      } else {
+        db.collection(resourceType).findOne({id: resourceId}, callback)
+      }
+    })
+  }
+
+  function resolveNestedProperties (resultList, includeReference, callback) {
+    const promises = []
+
+    // resolve references in nested objects recursively
+    const evaluateProperties = (property, nextProperty) => {
+      if (property == null) {
+        return callback(new Error('Undefined resource property.'))
+      }
+
+      if (Array.isArray(property)) {
+        for (var n = 0; n < property.length; n++) {
+          if (nextProperty[0] && property[n][nextProperty[0]] != null) {
+            evaluateProperties(property[n][nextProperty[0]], nextProperty.slice(1))
+          } else if (property[n].reference) {
+            promises.push(new Promise((resolve, reject) => {
+              resolveReference(property[n].reference, (err, data) => {
+                if (err) {
+                  reject(err)
+                } else {
+                  resolve(data)
+                }
+              })
+            }))
+          } else {
+            return callback(new Error(`Invalid resource reference: "${JSON.stringify(property[n])}"`))
+          }
+        }
+      } else {
+        if (nextProperty[0] && property[nextProperty[0]] != null) {
+          evaluateProperties(property[nextProperty[0]], nextProperty.slice(1))
+        } else if (property.reference) {
+          promises.push(new Promise((resolve, reject) => {
+            resolveReference(property.reference, (err, data) => {
+              if (err) {
+                reject(err)
+              } else {
+                resolve(data)
+              }
+            })
+          }))
+        } else {
+          return callback(`Invalid resource reference: "${JSON.stringify(property)}"`)
+        }
+      }
+    }
+
+    evaluateProperties(resultList[includeReference[0]], includeReference.slice(1))
+
+    // resolve promises
+    Promise.all(promises)
+      .then((res) => {
+        callback(null, res)
+      }).catch((err) => {
+        callback(err)
+      })
+  }
+
+  const parseSearchExpression = (exp, callback) => {
+    const itemList = exp.split(':')
```

Comment:
Check the length of the split result to see if it's long enought or `itemList[1].split('.')` will throw an exception and not call your callback.